### ISSUE TITLE
Cannot upgrade production cluster from 17.09 to 18.03 - STUCK!

### ISSUE THREAD

**Issue by michaelkrog** [moby/moby#36712]
**Description**
When upgrading our production cluster, one of our managers were unable to reconnect. I demoted it and tried to join again. But then the following happens:

```
docker swarm join --token SWMTKN-1-3q0ksqtg0yh7xdh1nu7aq2o827eysu3hrj6ngeyuwcrsd0kg44-4581tn4pfd8pjlju85h5igzn1 10.201.0.13:2377

Error response from daemon: Timeout was reached before node joined. The attempt to join the swarm will continue in the background. Use the "docker info" command to see the current swarm status of your node.
```

When on the current (still to be upgraded leader), this is the output of `docker node ls`:
```
ID                            HOSTNAME              STATUS              AVAILABILITY        MANAGER STATUS
yec6palr4x5up62jcjqe083v9 *   engine01.codezoo.dk   Ready               Active              Leader
ez1g2rn72dau4vlwf6c86mwod     engine02.codezoo.dk   Ready               Active              Unreachable
o2ie10iy3yt30jmubfhyw1kpi     engine03.codezoo.dk   Ready               Active              Reachable
ev6rqsnbbqe8n0x97cw0jzed4     engine04.codezoo.dk   Ready               Active              
ok647s8c0s6nrn5fbklyu0l1h     engine05.codezoo.dk   Ready               Active   
```

When on another (upgraded) leader, this is the output of `docker node ls`:
```
ID                            HOSTNAME              STATUS              AVAILABILITY        MANAGER STATUS      ENGINE VERSION
yec6palr4x5up62jcjqe083v9     engine01.codezoo.dk   Ready               Active              Leader              17.09.0-ce
ez1g2rn72dau4vlwf6c86mwod     engine02.codezoo.dk   Ready               Active              Unreachable         18.03.0-ce
o2ie10iy3yt30jmubfhyw1kpi *   engine03.codezoo.dk   Ready               Active              Reachable           18.03.0-ce
ev6rqsnbbqe8n0x97cw0jzed4     engine04.codezoo.dk   Ready               Active                                  18.03.0-ce
ok647s8c0s6nrn5fbklyu0l1h     engine05.codezoo.dk   Ready               Active                                  18.03.0-ce
```

**Steps to reproduce the issue:**
1. Create a cluster(3 manager + 2 nodes) with 17.09
2. Upgrade it to 18.03

**Describe the results you received:**
Error response from daemon: This node is not a swarm manager. Use "docker swarm init" or "docker swarm join" to connect this node to swarm and try again.

**Describe the results you expected:**
Node joined swarm

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:18 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:40:56 2017
 OS/Arch:      linux/amd64
 Experimental: false

and

Client:
 Version:	18.03.0-ce
 API version:	1.37
 Go version:	go1.9.4
 Git commit:	0520e24
 Built:	Wed Mar 21 23:10:01 2018
 OS/Arch:	linux/amd64
 Experimental:	false
 Orchestrator:	swarm

Server:
 Engine:
  Version:	18.03.0-ce
  API version:	1.37 (minimum version 1.12)
  Go version:	go1.9.4
  Git commit:	0520e24
  Built:	Wed Mar 21 23:08:31 2018
  OS/Arch:	linux/amd64
  Experimental:	false
```

**Output of `docker info`:**

```
Containers: 907
 Running: 30
 Paused: 0
 Stopped: 877
Images: 236
Server Version: 17.09.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 2348
 Dirperm1 Supported: true
Logging Driver: gelf
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: active
 NodeID: yec6palr4x5up62jcjqe083v9
 Is Manager: true
 ClusterID: e20qw9yysm9syzh3us19idl8z
 Managers: 3
 Nodes: 5
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
  Force Rotate: 0
 Autolock Managers: false
 Root Rotation In Progress: false
 Node Address: 10.201.0.11
 Manager Addresses:
  10.201.0.11:2377
  10.201.0.12:2377
  10.201.0.13:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-116-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.65GiB
Name: engine01.codezoo.dk
ID: SIRC:Z5VT:MSC5:UVOO:QWD5:RAHK:U3P2:HNXM:EM7Z:THIN:SC2R:5LUO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support


and 


Client:
 Version:	18.03.0-ce
 API version:	1.37
 Go version:	go1.9.4
 Git commit:	0520e24
 Built:	Wed Mar 21 23:10:01 2018
 OS/Arch:	linux/amd64
 Experimental:	false
 Orchestrator:	swarm

Server:
 Engine:
  Version:	18.03.0-ce
  API version:	1.37 (minimum version 1.12)
  Go version:	go1.9.4
  Git commit:	0520e24
  Built:	Wed Mar 21 23:08:31 2018
  OS/Arch:	linux/amd64
  Experimental:	false
michael@engine02:~$ docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 205
Server Version: 18.03.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 446
 Dirperm1 Supported: true
Logging Driver: gelf
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: error
 NodeID: 
 Error: rpc error: code = DeadlineExceeded desc = context deadline exceeded
 Is Manager: false
 Node Address: 10.201.0.12
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfd04396dc68220d1cecbe686a6cc3aa5ce3667c
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-116-generic
Operating System: Ubuntu 16.04.4 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 15.65GiB
Name: engine02.codezoo.dk
ID: QLO7:YKVQ:GICP:V4F3:3EE5:JOQC:OQFM:4YPG:4XFF:HKIY:NVCI:CNZ3
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Ubuntu 16.04 on vmware.

**Comment by michaelkrog** [comment]
And following the disaster recovery her: https://docs.docker.com/engine/swarm/admin_guide/#recover-from-disaster

...simply caused us to loose our whole production setup..  

I think my job down the toilet!!!

**Comment by cpuguy83** [comment]
Can you provide some more details?
Logs from the "stuck" daemon and the manager node would be good.
Also stack dumps from both of these nodes. You can get this by doing `curl --unix-socket /var/run/docker.sock http://./debug/pprof/goroutine?debug=2` on the node.

**Comment by cpuguy83** [comment]
When you say you followed those steps, what does that mean exactly? What steps did you perform on what node?

**Comment by michaelkrog** [comment]
On the remaining leader(engine01) I backed up the /var/lib/docker/swarm folder and then upgraded to 18.03, hoping having all nodes on the same version fix it.

But now that node cannot start up...
```
$ curl --unix-socket /var/run/docker.sock http://./debug/pprof/goroutine?debug=2
curl: (7) Couldn't connect to server

$ sudo service docker restart
Job for docker.service failed because the control process exited with error code. See "systemctl status docker.service" and "journalctl -xe" for details.

$ sudo journalctl -xe
-- Subject: Unit user@1006.service has begun shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit user@1006.service has begun shutting down.
Mar 27 21:11:45 engine01.codezoo.dk systemd[3266]: Stopped target Default.
-- Subject: Unit UNIT has finished shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit UNIT has finished shutting down.
Mar 27 21:11:45 engine01.codezoo.dk systemd[3266]: Reached target Shutdown.
-- Subject: Unit UNIT has finished start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit UNIT has finished starting up.
-- 
-- The start-up result is done.
Mar 27 21:11:45 engine01.codezoo.dk systemd[3266]: Stopped target Basic System.
-- Subject: Unit UNIT has finished shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit UNIT has finished shutting down.
Mar 27 21:11:45 engine01.codezoo.dk systemd[3266]: Stopped target Timers.
-- Subject: Unit UNIT has finished shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit UNIT has finished shutting down.
Mar 27 21:11:45 engine01.codezoo.dk systemd[3266]: Stopped target Sockets.
-- Subject: Unit UNIT has finished shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit UNIT has finished shutting down.
Mar 27 21:11:45 engine01.codezoo.dk systemd[3266]: Stopped target Paths.
-- Subject: Unit UNIT has finished shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit UNIT has finished shutting down.
Mar 27 21:11:45 engine01.codezoo.dk systemd[3266]: Starting Exit the Session...
-- Subject: Unit UNIT has begun start-up
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit UNIT has begun starting up.
Mar 27 21:11:45 engine01.codezoo.dk systemd[3266]: Received SIGRTMIN+24 from PID 3306 (kill).
Mar 27 21:11:45 engine01.codezoo.dk systemd[3267]: pam_unix(systemd-user:session): session closed for user prtg
Mar 27 21:11:45 engine01.codezoo.dk systemd[1]: Stopped User Manager for UID 1006.
-- Subject: Unit user@1006.service has finished shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit user@1006.service has finished shutting down.
Mar 27 21:11:45 engine01.codezoo.dk systemd[1]: Removed slice User Slice of prtg.
-- Subject: Unit user-1006.slice has finished shutting down
-- Defined-By: systemd
-- Support: http://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit user-1006.slice has finished shutting down.
Mar 27 21:12:13 engine01.codezoo.dk sudo[3326]:  michael : TTY=pts/0 ; PWD=/home/michael ; USER=root ; COMMAND=/bin/journalctl -xe
Mar 27 21:12:13 engine01.codezoo.dk sudo[3326]: pam_unix(sudo:session): session opened for user root by michael(uid=0)
```

**Comment by cpuguy83** [comment]
How about `journalctl -fu docker`

**Comment by cpuguy83** [comment]
@michaelkrog This is because `/etc/docker/daemon.json` now conflicts with the systemd unit file.
You'll need to add a systemd drop-in file to overwrite the `ExecStart` command.

Basically, place this in `/etc/systemd/system/docker.service.d/start.conf`

```
[Service]
ExecStart=/usr/bin/dockerd
```

Then docker should start correctly.

**Comment by michaelkrog** [comment]
Super.. I changed that..

Still, when I restore the /var/lib/docker/swarm directory with the contents of the backup., I have no success. 

I get this: `The swarm does not have a leader. It's possible that too few managers are online. Make sure more than half of the managers are online.`

**Comment by cpuguy83** [comment]
At that point quorum has been lost. I'm not sure what else may have been done here, but if you are unable to get at least 2 managers online (and have 3 actually registered) then you'll have to follow the directions for loss of quorum.

**Comment by michaelkrog** [comment]
Ok.. I have copy of the swarm folder from the leader.

Is it possible to the extract the info, so I can begin set it up manually again?

**Comment by cpuguy83** [comment]
`docker swarm init --force-new-cluster` does this.
It creates a new cluster from the existing state, but without other nodes.

**Comment by michaelkrog** [comment]
Just to follow up on this.

I tried to follow the steps for [backup and disaster recovery](https://docs.docker.com/datacenter/ucp/1.1/high-availability/backups-and-disaster-recovery/#backup-policy) and initing a new cluster.

But the master I got the backup from could not start up when I copied the data in. It just hang on `sudo service docker start`.

I copied data into one of the other masters, and it started fine, but it showed no data afterwards.

I was, of course, under pressure when all this happened so I might have made mistakes. I the end we had to recreate our whole production environement.

Thank you for trying to help me out though. Much appreciated.
